src/hotspot/share/opto/compile.cpp

```diff
@@ -1852,6 +1855,90 @@ void Compile::process_for_post_loop_opts_igvn(PhaseIterGVN& igvn) {
   }
 }
 
+void Compile::record_unstable_if(UnstableIfTrap* trap) {
+  if (AggressiveLivenessForUnstableIf) {
+    _unstable_ifs.append(trap);
+  }
+}
+
+void Compile::invalidate_unstable_if(CallStaticJavaNode* unc) {
+  for (int i = 0; i < _unstable_ifs.length(); ++i) {
+    UnstableIfTrap* trap = _unstable_ifs.at(i);
+    if (trap->uncommon_trap() == unc) {
+      trap->set_bci(-1);
+      break;
+    }
+  }
+}
+
+uint trivial_unstable_ifs          = 0;
+uint unstable_ifs_all              = 0;
+
+void Compile::preprocess_unstable_ifs() {
```

Comment:
> Yes, this is for 8287385. I will remove preprocessing logic from this PR.

Okay.

> Because we can determine a unstable_if trap is trivial after parsing. My idea is to do the leftover parsing job in this preprocess function. That's why I think preprocess should before `PhaseRemoveUseless`.

Then you should consider my proposal about cleaning the list and update counters in `Compile::remove_useless_nodes()`. Let discuss it in next changes.

Reply:
I move preprocess after PhaseRemoveUseless phase. I think it's reasonable to count only useful nodes.